pkg/async/schedule/flytescheduler/workflow_executor.go

```diff
@@ -0,0 +1,234 @@
+	schedulerEpochTime := workflowExecConfig.SchedulerEpochTime
+
+	var fromTime time.Time
+	// If the scheduler epoch time ise set then use that to catch up till the current time
+	// else check the last exec time for the schedule and if that also is not set then use current time with
+	// configured jitter
+	if !schedulerEpochTime.IsZero() {
+		fromTime = schedulerEpochTime
+	} else {
+		if !lastExecTime.IsZero() {
+			fromTime = lastExecTime
+		} else {
+			fromTime = time.Now().Add(-jitter)
+		}
+	}
+
+	var catchUpTimes []time.Time
+	var err error
+	catchUpTimes, err = w.goGfInterface.GetCatchUpTimes(s, fromTime, toTime)
+	if err != nil {
+		return err
+	}
+	var catchupTime time.Time
+	for _, catchupTime = range catchUpTimes {
+		err := fire(ctx, w.executionManager, catchupTime, s)
+		if err != nil {
+			logger.Errorf(ctx, "unable to fire the schedule %v at %v time due to %v", s, catchupTime, err)
+		} else {
+			w.snapshot.UpdateLastExecutionTime(nameOfSchedule, catchupTime)
+		}
+	}
+
+	return nil
+}
+
+func fire(ctx context.Context, executionManager mgInterfaces.ExecutionInterface, scheduledTime time.Time,
+	s models.SchedulableEntity) error {
+
+	literalsInputMap := map[string]*core.Literal{}
+	literalsInputMap[s.KickoffTimeInputArg] = &core.Literal{
+		Value: &core.Literal_Scalar{
+			Scalar: &core.Scalar{
+				Value: &core.Scalar_Primitive{
+					Primitive: &core.Primitive{
+						Value: &core.Primitive_Datetime{
+							Datetime: timestamppb.New(scheduledTime),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	executionRequest := admin.ExecutionCreateRequest{
+		Project: s.Project,
+		Domain:  s.Domain,
+		Name:    "f" + strings.ReplaceAll(uuid.New().String(), "-", "")[:19],
+		Spec: &admin.ExecutionSpec{
+			LaunchPlan: &core.Identifier{
+				ResourceType: core.ResourceType_LAUNCH_PLAN,
+				Project:      s.Project,
+				Domain:       s.Domain,
+				Name:         s.Name,
+				Version:      s.Version,
+			},
+			Metadata: &admin.ExecutionMetadata{
+				Mode:        admin.ExecutionMetadata_SCHEDULED,
+				ScheduledAt: timestamppb.New(scheduledTime),
+			},
+			// No dynamic notifications are configured either.
+		},
+		// No additional inputs beyond the to-be-filled-out kickoff time arg are specified.
+		Inputs: &core.LiteralMap{
+			Literals: literalsInputMap,
+		},
+	}
+
+	_, err := executionManager.CreateExecution(context.Background(), executionRequest, scheduledTime)
+	if err != nil {
+		logger.Error(ctx, "failed to create  execution create  request due to %v", err)
+		return err
+	}
+	return nil
+}
+
+func (w *workflowExecutor) Run() {
+	ctx := context.Background()
+	c, err := w.db.SchedulableEntityRepo().GetAll(ctx)
+	if err != nil {
+		panic(fmt.Errorf("unable to run the workflow executor after reading the schedules due to %v", err))
+	}
+	// Start the go routine to write the snapshot map to the configured snapshot file
+	go w.CheckPointState(ctx)
+	catchUpTill := time.Now()
+	err = w.CatchUpAllSchedules(ctx, c.Entities, catchUpTill)
```

Comment:
This can spawn go routines later for each schedule .Right now this is synchronous